internal/metrics/metrics.go

```diff
@@ -83,111 +82,139 @@ func NewInformerMetrics(label string) *InformerMetrics {
+			Name: "principal_errors",
+			Help: "The total number of errors occurred in principal",
+		}, []string{"resource_type"}),
 	}
 }
 
-func NewAppProjectWatcherMetrics() *AppProjectWatcherMetrics {
-	am := &AppProjectWatcherMetrics{
-		ProjectsWatched: promauto.NewGauge(prometheus.GaugeOpts{
-			Name: "argocd_agent_watcher_appprojects_watched",
-			Help: "The total number of AppProjects watched by the agent",
-		}),
-		ProjectsAdded: promauto.NewCounter(prometheus.CounterOpts{
-			Name: "argocd_agent_watcher_appprojects_added",
-			Help: "The number of AppProjects that have been added to the agent",
-		}),
-		ProjectsUpdated: promauto.NewCounter(prometheus.CounterOpts{
-			Name: "argocd_agent_watcher_appprojects_updated",
-			Help: "The number of AppProjects that have been updated",
+func NewAgentMetrics() *AgentMetrics {
+	return &AgentMetrics{
+		EventReceived: promauto.NewCounter(prometheus.CounterOpts{
+			Name: "agent_events_received",
+			Help: "The total number of events received by agent",
 		}),
-		ProjectsRemoved: promauto.NewCounter(prometheus.CounterOpts{
-			Name: "argocd_agent_watcher_appprojects_removed",
-			Help: "The number of AppProjects that have been removed from the agent",
+		EventSent: promauto.NewCounter(prometheus.CounterOpts{
+			Name: "agent_events_sent",
+			Help: "The total number of events sent by agent",
 		}),
+
+		EventProcessingTime: promauto.NewHistogramVec(prometheus.HistogramOpts{
+			Name: "agent_event_processing_time",
+			Help: "Histogram of time taken to process events (in seconds)",
+		}, []string{"status", "agent_mode", "resource_type"}),
+
+		AgentErrors: promauto.NewCounterVec(prometheus.CounterOpts{
+			Name: "agent_errors",
+			Help: "The total number of errors occurred in agent",
+		}, []string{"resource_type"}),
 	}
-	return am
 }
 
-// func (am *ApplicationWatcherMetrics) SetWatched(num int64) {
-// 	am.AppsWatched.Set(float64(num))
-// }
+// AvgCalculationInterval is time interval for agent connection time calculation
+var AvgCalculationInterval = 3 * time.Minute
 
-// func (am *ApplicationWatcherMetrics) AppAdded() {
-// 	am.AppsWatched.Inc()
-// 	am.AppsAdded.Inc()
-// }
+// ConnectionTimeMap is an utility for AvgAgentConnectionTime
+// var ConnectionTimeMap = make(map[string]time.Time)
+type AgentConnectionTime struct {
+	Lock sync.RWMutex
 
-// func (am *ApplicationWatcherMetrics) AppRemoved() {
-// 	am.AppsWatched.Dec()
-// 	am.AppsRemoved.Inc()
-// }
+	// key: AgentName
+	// value: connection time of agent
+	// - acquire 'lock' before accessing
+	ConnectionTimeMap map[string]time.Time
+}
+
+var agentConnectionTime = &AgentConnectionTime{
+	ConnectionTimeMap: make(map[string]time.Time),
+}
 
-// func (am *ApplicationWatcherMetrics) AppUpdated() {
-// 	am.AppsUpdated.Inc()
-// }
+// SetAgentConnectionTime inserts connection time of new agent
+func SetAgentConnectionTime(agentName string, start time.Time) {
+	agentConnectionTime.Lock.RLock()
+	defer agentConnectionTime.Lock.RUnlock()
```

Comment:
Should use a write lock, not a read lock, when accessing maps for writing

Reply:
Ah :(, fixed it.